docs/src/proposals/stake-v2.md

```diff
@@ -0,0 +1,211 @@
+---
+title: Stake Account Version 2
+---
+
+The current stake program is battle-tested, but time has revealed a few important
+issues:
+
+1. Redelegations are not possible
+
+Whales can't easily redelegate their positions, since they do not want to lose
+out on an epoch of rewards.
+
+2. Rewards calculations at epoch boundaries take too long
+
+Since all of the stake accounts simultaneously receive rewards at the start of an epoch,
+the network experiences a big load to process these accounts. This feature
+is extremely useful for users, since they don't need to do anything to claim
+rewards. On the flip side, it introduces too much load on validators.
+
+With a new stake account, we can address both of these issues.
+
+## New Stake Account
+
+The current `StakeState` is 200 bytes, and can be in 4 states:
+
+* `Uninitialized`
+* `Initialized(Meta)`: the rent-exempt reserve, authorized keys, and lockup
+* `Stake(Meta, Stake)`: `Delegation` (validator, stake amount, activation / deactivation epochs, warmup rate) + `credits_observed` (used for rewards calc)
+* `RewardsPool`: deprecated (?)
+
+To properly perform a redelegation, we need to store at least the
+next validator's vote pubkey and a redelegation epoch, which means that
+redelegation absolutely requires a larger struct.
+
+### Structure
+
+We need to introduce a new type and an upgrade path from legacy / V1 stakes:
+
+```
+struct StakeV2 {
+    meta: Meta,
+    stake: Stake,
+    last_claim_epoch: Epoch, // needed to know if it's up-to-date, see `ClaimRewards`
+    redelegation: Redelegation,
+}
+struct Redelegation {
```

Comment:
I've been sniffing around the stake program and runtime, considering how to add redelegation support with the lowest risk possible, and the best I've come up with is a new entry to `StakeState` that looks like `StakeV2(Meta, Stake, Stake),`

With two `Stake` structs, the source and destination for a redelegation operation, I think we can teach the runtime to process each of them basically independently and the math for stake weights+rewards should all work out with very minimal changes, provided that the stake program sets the `Stake` struct values correctly.  Redelegations that support multiple epochs should also be handled.

I have a bit more prototyping work I want to do soon to continue exploring this approach

Reply:
Hehe, that's a slick way of doing it.  On the flipside, I worry about encoding the delegation in two places and shifting it over bit by bit, since it could be more error-prone than calculating everything based on redelegation epoch, current epoch, stake history, and total delegation. That can give us the amount delegated to each validator consistently every time.

Reply:
I'm pretty sure it'll just work if the two `Delegation` structs are correctly setup by the stake program on a redelegation for deactivation/activation on the same epoch.  But yeah, that's the "bit more prototyping work I want to do soon" bit talking again

Reply:
After some more reflection, I like your approach more.  Proposal updated!